### PR TITLE
Replace sprintf with snprintf in the runtime

### PR DESCRIPTION
**Pull request by jhh67**
Replace `sprintf` with `snprintf` in the runtime

Resolves https://github.com/Cray/chapel-private/issues/3996.

### CODE REVIEW

File: runtime/src/chpl-launcher-common.c [hunk 1]
```diff
@@ -303,9 +306,10 @@ void chpl_launcher_record_env_var(const char* evName, const char *evVal) {
   evList = chpl_mem_realloc(evList, evListSize,
                             CHPL_RT_MD_COMMAND_BUFFER, -1, 0);
   if (evListSizeWas == 0) {
-    sprintf(evList, "%s=%s", evName, evVal);
+    snprintf(evList, evListSize, "%s=%s", evName, evVal);
```
Comment (riftEmber) [inline runtime/src/chpl-launcher-common.c:309]: here

File: runtime/src/chpl-launcher-common.c [hunk 2]
```diff
@@ -303,9 +306,10 @@ void chpl_launcher_record_env_var(const char* evName, const char *evVal) {
...
   } else {
-    sprintf(evList + evListSizeWas - 1, " %s=%s", evName, evVal);
+    snprintf(evList + evListSizeWas - 1, evListSize - (evListSizeWas - 1),
```
Comment (riftEmber) [inline runtime/src/chpl-launcher-common.c:311]: here

File: runtime/src/chpl-string-support.c [hunk 3]
```diff
@@ -178,7 +178,7 @@ string_index(c_string x, int i, int32_t lineno, int32_t filename) {
   }
   buffer = chpl_mem_allocMany(1, 2, CHPL_RT_MD_STR_COPY_DATA,
                               lineno, filename);
-  sprintf(buffer, "%c", x[i-1]);
+  snprintf(buffer, 2, "%c", x[i-1]);
```
Comment (riftEmber) [inline runtime/src/chpl-string-support.c:181]: here (maybe this one doesn't count as it's a literal? not sure)

File: runtime/src/chplmemtrack.c [hunk 4]
```diff
@@ -173,8 +173,9 @@ void chpl_setMemFlags(void) {
     if (chpl_numNodes == 1) {
       memLogFile = fopen(memLog, "w");
     } else {
-      char* filename = (char*)sys_malloc((strlen(memLog)+10)*sizeof(char));
-      sprintf(filename, "%s.%" PRI_c_nodeid_t, memLog, chpl_nodeID);
+      int size = (strlen(memLog)+10)*sizeof(char);
+      char* filename = (char*)sys_malloc(size);
+      snprintf(filename, size, "%s.%" PRI_c_nodeid_t, memLog, chpl_nodeID);
```
Comment (riftEmber) [inline runtime/src/chplmemtrack.c:178]: here, although we multiply by `sizeof(char)` when declaring `size` it it declared as an `int`

File: runtime/src/chplmemtrack.c [hunk 5]
```diff
@@ -602,15 +603,17 @@ printMemAllocs(chpl_mem_descInt_t description, int64_t threshold,
   }
   qsort(table, n, sizeof(memTableEntry*), descCmp);
 
-  loc = (char*)sys_malloc((filenameWidth+numberWidth+1)*sizeof(char));
+  int locSize = (filenameWidth+numberWidth+1)*sizeof(char);
```
Comment (riftEmber) [inline runtime/src/chplmemtrack.c:606]: here, for use in both `sys_malloc` and `snprintf` below

File: runtime/src/launch/pbs-aprun/launch-pbs-aprun.c [hunk 6]
```diff
@@ -116,7 +117,8 @@ static char* genQsubOptions(char* genFilename, char* projectString, qsubVersion
 
   if (generate_qsub_script) {
     pid_t mypid = debug ? 0 : getpid();
-    sprintf(qsubFilename, "qsub.%s-%d", genFilename, (int) mypid);
+    snprintf(qsubFilename, qsubFilenameSize, "qsub.%s-%d", genFilename,
```
Comment (riftEmber) [inline runtime/src/launch/pbs-aprun/launch-pbs-aprun.c:120]: here (`size_t` stored into an `int`)

File: runtime/src/launch/slurm-gasnetrun_common/slurm-gasnetrun_common.h [hunk 7]
```diff
@@ -158,7 +158,7 @@ static int propagate_environment(char* buf)
   // to leave out something important.
   char *enviro_keys = chpl_get_enviro_keys(',');
   if (enviro_keys)
-    len += sprintf(buf, " -E '%s'", enviro_keys);
+    len += snprintf(buf, size, " -E '%s'", enviro_keys);
```
Comment (riftEmber) [inline runtime/src/launch/slurm-gasnetrun_common/slurm-gasnetrun_common.h:161]: here

File: runtime/src/launch/slurm-srun/launch-slurm-srun.c [hunk 8]
```diff
@@ -390,76 +394,78 @@ static char* chpl_launch_create_command(int argc, char* argv[],
...
     // launch the job using srun
-    sprintf(baseCommand, "srun %s", iCom);
+    snprintf(baseCommand, sizeof(baseCommand), "srun %s", iCom);
   }
 
   // copy baseCommand into command and return it
   size = strlen(baseCommand) + 1;
   command = chpl_mem_allocMany(size, sizeof(char), CHPL_RT_MD_COMMAND_BUFFER, -1, 0);
-  sprintf(command, "%s", baseCommand);
+  snprintf(command, size, "%s", baseCommand);
```
Comment (riftEmber) [inline runtime/src/launch/slurm-srun/launch-slurm-srun.c:468]: here

**Review by riftEmber — CHANGES_REQUESTED** [review]
For consistency and to future proof against warnings/strict standards, I think we should stick to feeding in a `size_t` to `snprintf`, either by casting or multiplying by `sizeof(char)` as appropriate. I've marked all occurences I found where it gets an `int` instead.

**Review by riftEmber — APPROVED** [review]
thanks!

**Comment by bradcray** [comment]
Thanks John!

**Pull request merged**